sqlmesh/utils/file.py

```diff
@@ -0,0 +1,129 @@
+        self._path = path
+        self._original_contents: t.Optional[bytes] = None
+        self._is_locked = False
+
+    @property
+    def path(self) -> str:
+        """The key which represents a path to a file."""
+        return self._path
+
+    @property
+    def lock_path(self) -> str:
+        """The key which represents a path to the file's lock."""
+        return f"{self.path}.lock"
+
+    def read(self) -> t.Optional[bytes]:
+        """Reads data from the file."""
+        if self._fs.exists(self.path):
+            if not self._is_locked:
+                ... # TODO: do whatever we think is right here
+            content = self._fs.cat(self.path)
+            self._original_contents = content[:]
+            return content
+        return None
+    
+    def write(self, content: bytes) -> None:
+        """Writes data within the transaction."""
+        if not self._is_locked:
+            # TODO: do whatever we think is right here
+            raise RuntimeError("Cannot write to a file without a lock.")
+        self._fs.pipe(self.path, content)
+
+    def rollback(self) -> None:
+        """Rolls back the transaction."""
+        if self._original_contents is not None:
+            self._fs.pipe(self.path, self._original_contents)
+
+    def acquire_lock(
+        self, blocking: bool = True, timeout: float = -1
+    ) -> bool:
+        """Acquires a lock on a key.
+        
+        Acquire a lock, blocking or non-blocking.
+
+        When invoked with the blocking argument set to True (the default), block until
+        the lock is unlocked, then set it to locked and return True.
+
+        When invoked with the blocking argument set to False, do not block. If a call
+        with blocking set to True would block, return False immediately; otherwise, set
+        the lock to locked and return True.
+
+        When invoked with the floating-point timeout argument set to a positive value,
+        block for at most the number of seconds specified by timeout and as long as the
+        lock cannot be acquired. A timeout argument of -1 specifies an unbounded wait.
+        If blocking is False, timeout is ignored. The stdlib would raise a ValueError.
+
+        The return value is True if the lock is acquired successfully, False if
+        not (for example if the timeout expired).
+        """
+        if self._is_locked:
+            # Make this idempotent for a single instance.
+            return True
+        lock_id = uuid4()
+        wait_time = 0
+        while self._fs.exists(self.lock_path):
+            if not blocking:
+                return False
+            time.sleep(1)
+            wait_time += 1
+            if timeout > 0 and wait_time > timeout:
+                return False
+        self._fs.pipe(self.lock_path, json.dumps(
+            # Add more fields here if needed.
+            {"gid": str(lock_id)},
+        ))
+        self._jitter()
```

Comment:
why does the jitter happen here? shouldn't this happen in the sleep on line 88?

Reply:
can we just use jitter instead of the hardcode 1 second sleep?

Reply:
> sleep on line 88
That sleep occurs when we know the file is locked and we are patiently waiting for the writer to finish.

This jitter occurs in the event there is no lock but 2 or more writers try to acquire a lock (writing a file). Most cloud storage providers have strong Read-after-write consistency so we want to jitter here so that there is a clear winner in the event of near-simultaneous writes before the read call. At least that was my thinking.

Reply:
> can we just use jitter instead of the hardcode 1 second sleep?

Yeah we can here too is fine actually

Reply:
sorry, i'm not sure i understand the nuance.

fs.pipe() is a blocking call and should be atomic right? why do we need a jitter before the fs.cat call?

Reply:
no lock exists, two writers

A, B both pipe simultaenously

will there be an error?

will A think it wrote but then b overwrites?

if a is overwritten, on cat it will see guid is not correct and be safe

the only issue is if  a writes and then cats and is correct, but then b overwrites, and then cats

Reply:
Yes that makes sense actually. I will just use the one jitter. in the waiter

Reply:
> the only issue is if a writes and then cats and is correct, but then b overwrites, and then cats

Yeah this is a better way to describe what I wanted to avoid, its pretty low probability but I figured the jitter didnt hurt there too

Reply:
is there any way to make sure this never happens?

Reply:
Yes the jitter makes it almost impossible because the window is microscopic to do read+write AFTER passing the if not exists check and AFTER a cat verifies the gid for another writer. Microscopic for sure

Reply:
s3 has a concept of tags, does fsspec have this?

Reply:
https://stackoverflow.com/a/75347123/2554662